### PR TITLE
Focused Launch: Fix billing period tracking in summary

### LINKED ISSUE #49713: Focused Launch - Summary View: Prices jump from monthly to yearly after selecting a free plan
If the user has a monthly plan selected in the summary view and then clicks on the free plan, the price of the plans above jump from monthly to annually while staying in the summary view.

#### Acceptance criteria
* [ ] The list of plans in the summary view (either monthly or annually) remains the same after selecting a free plan.

#### Expected behavior
When the user selects a Free plan after selecting a Monthly plan, I expect the prices of the paid plans in the summary view not to change (to annually billed plans). 

#### What actually happened
https://www.loom.com/share/611f316948594d3e894f1d6b7ed997d2

### PR DESCRIPTION
**Pull request by alshakero**
#### Changes proposed in this Pull Request

* I added `getLastPlanBillingPeriod` selector to the plan store to get the billing period of the current or last selected paid plan's period.
* I added a private (i.e unexported) action creator `__setBillingPeriod` to the launch store. And I changed the `setPlanProductId` action creator into a generator that queries the plans store for the plans billing period using its productId, then it dispatches the billing period using the newly-added `__setBillingPeriod` and the reducer records it. The makes the plan readonly and can only be changed by changing the plan. 
* I updated the `usePlans` hook to return the plan products, too.
* I refactored summary view logic around plans and plan products.

#### Testing instructions

1. Create or reuse a site that is created with `/start` flow.
2. Sandbox this site.
3. In `apps/editing-toolkit` run `yarn dev --sync`.
4. Go to https://[yoursite].wordpress.com/wp-admin/post-new.php
5. Click launch.
6. Go to detailed plan page, pick a monthly plan.
7. In summary, now pick a free plan, the prices should remain monthly.
8. Pick a non-default paid plan, it should be rendered in summary, re-pick the default paid plan (Premium), the non-default plan should remain listed.

Fixes #49713

### CODE REVIEW

**Review by ciampo — CHANGES_REQUESTED** [review]
> The launch store had a selector that returns the current paid plan. While it is meant to return the current or once-picked paid plan. I fixed that selector.

I wasn't really aware of how that selector was meant to work until I read your explanation.

As a user of the Launch data-store, when seeing a selector called `getPaidPlanProductId()`, I would expect the result of that selector to be `undefined` if there isn't a selected paid plan.

The fact that [the result is assigned to a variable named `selectedPaidPlanProductId`](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/launch/src/focused-launch/summary/index.tsx#L290) would make me even more confused. If the user selects the free plan, why would a variable called `selectedPaidPlanProductId` still hold the value of a paid plan that is not currently selected?

---

To bring clarity, I think we should rename a few variables (sorry for the added verbosity, but it would make the code much easier to understand):
- the [reducer](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/data-stores/src/launch/reducer.ts#L67-L71) could be called `lastSelectedPaidPlanProductId`
- the [selector](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/data-stores/src/launch/selectors.ts#L30-L39) could be called `getLastSelectedPaidPlanProductId()`
- this [variable](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/launch/src/focused-launch/summary/index.tsx#L290) could be called `lastSelectedPaidPlanProductId`
- this [variable](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/launch/src/focused-launch/summary/index.tsx#L299) could be called `lastSelectedPaidPlan`

Finally, I believe that we should refactor [this logic](https://github.com/Automattic/wp-calypso/blob/86b20569c9836bab85b59c71778bef5e8c0cc54a/packages/launch/src/focused-launch/summary/index.tsx#L307-L318) too, because now `selectedPaidPlan` can be defined even when the user hasn't currently selected a paid plan (we could use `selectedPlan` instead, with an additional check to see it's a free plan or a paid plan?)

**Comment by alshakero** [comment]
@ciampo I agree it was pretty confusing. I removed the reducer, the selector, and moved the logic to the focused launch.

File: packages/launch/src/hooks/use-plans.tsx [hunk 1]
```diff
@@ -15,21 +15,35 @@ import LaunchContext from '../context';
...
+		const defaultFreePlan = select( PLANS_STORE ).getDefaultFreePlan( locale );
+		const defaultPaidPlan = select( PLANS_STORE ).getDefaultPaidPlan( locale );
+		const defaultPaidPlanProduct = select( PLANS_STORE ).getPlanProduct(
+			defaultPaidPlan?.periodAgnosticSlug,
+			billingPeriod
+		);
+		const defaultFreePlanProduct = select( PLANS_STORE ).getPlanProduct(
+			defaultFreePlan?.periodAgnosticSlug,
+			billingPeriod
+		);
```
Comment (StefanNieuwenhuis) [inline packages/launch/src/hooks/use-plans.tsx]: I think this is a bit cleaner and easier to maintain. What do you think?

```suggestion
		const plansStore = select(PLANS_STORE);
		const defaultFreePlan = plansStore.getDefaultFreePlan( locale );
		const defaultPaidPlan = plansStore.getDefaultPaidPlan( locale );
		const defaultPaidPlanProduct = plansStore.getPlanProduct(
			defaultPaidPlan?.periodAgnosticSlug,
			billingPeriod
		);
		const defaultFreePlanProduct = plansStore.getPlanProduct(
			defaultFreePlan?.periodAgnosticSlug,
			billingPeriod
		);
```

**Review by StefanNieuwenhuis — COMMENTED** [review]
The behavior looks good to me (see https://www.loom.com/share/f637c03af41d4b14abf78f1375afcfa6) and can you address my comment, please?

File: packages/launch/src/hooks/use-plans.tsx (hunk 1 above)
Comment (razvanpapadopol) [reply packages/launch/src/hooks/use-plans.tsx]: Also, don't forget to add the second argument to `useSelect` function which in this case should be `[locale, billingPeriod]`. See p7H4VZ-2Q6-p2#comment-9214

File: packages/launch/src/focused-launch/summary/index.tsx [hunk 2]
```diff
@@ -32,13 +32,12 @@ import FocusedLaunchSummaryItem, {
 	LeadingContentSide,
 	TrailingContentSide,
 } from './focused-launch-summary-item';
-import { LAUNCH_STORE, SITE_STORE, Plan } from '../../stores';
+import { LAUNCH_STORE, SITE_STORE, Plan, PlanProduct, PLANS_STORE } from '../../stores';
```
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/summary/index.tsx]: I think it's also fine to add on another line: `import type { Plan, PlanProduct } from '../../stores';`

File: packages/data-stores/src/launch/selectors.ts [hunk 3]
```diff
@@ -27,20 +27,6 @@ export const getSelectedDomain = ( state: State ): DomainSuggestions.DomainSugge
...
- *
- * @param state State
- */
-export const getPaidPlanProductId = ( state: State ): number | undefined => {
-	const productId = state.planProductId;
-	const isFree = select( PLANS_STORE ).isPlanProductFree( productId );
-
-	return productId && ! isFree ? state.planProductId : undefined;
-};
-
```
Comment (razvanpapadopol) [inline packages/data-stores/src/launch/selectors.ts:43]: I agree we don't need this selector as it is but if you look at changes from https://github.com/Automattic/wp-calypso/pull/48452 we are using twice this selector but just to check if the selected plan is a paid plan so `isSelectedPlanPaid` might be a good candidate for a selector on Launch store.

**Comment by alshakero** [comment]
Thanks everyone! I addressed all feedback.

**Review by StefanNieuwenhuis — APPROVED** [review]
LGTM! 🚢

File: packages/data-stores/src/launch/selectors.ts [hunk 4]
```diff
@@ -27,19 +27,9 @@ export const getSelectedDomain = ( state: State ): DomainSuggestions.DomainSugge
...
- * @param state State
- */
-export const getPaidPlanProductId = ( state: State ): number | undefined => {
-	const productId = state.planProductId;
-	const isFree = select( PLANS_STORE ).isPlanProductFree( productId );
-
-	return productId && ! isFree ? state.planProductId : undefined;
-};
+export const isSelectedPlanPaid = ( state: State ): boolean =>
+	Boolean( state.planProductId ) &&
```
Comment (ciampo) [inline packages/data-stores/src/launch/selectors.ts]: Bit of a nit, should we have a more explicit check instead of casting to boolean? It would make the code easier to understand (plus,  _techincally_, `planProductId` is a number and it could hold a value of `0`, which would return `false` when cast to Boolean)


```suggestion
	typeof state.planProductId !== 'undefined' &&
```

File: packages/launch/src/focused-launch/summary/index.tsx [hunk 5]
```diff
@@ -288,35 +288,46 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
...
 
-	const { defaultPaidPlan, defaultFreePlan } = usePlans();
-
 	// persist non-default selected paid plan if it's paid in order to keep displaying it in the plan picker
 	const [ nonDefaultPaidPlan, setNonDefaultPaidPlan ] = React.useState< Plan | undefined >();
+	const [ nonDefaultPaidPlanProduct, setNonDefaultPaidPlanProduct ] = React.useState<
+		PlanProduct | undefined
+	>();
+
+	const [ billingPeriod, setBillingPeriod ] = React.useState< PlanProduct[ 'billingPeriod' ] >(
```
Comment (ciampo) [inline packages/launch/src/focused-launch/summary/index.tsx]: Another nit: By adding this import to the file:

```typescript
import type { Plans } from '@automattic/data-stores';
```

we can then use the `PlanBillingPeriod` type:


```suggestion
	const [ billingPeriod, setBillingPeriod ] = React.useState< Plans.PlanBillingPeriod >(
```

File: packages/launch/src/focused-launch/summary/index.tsx [hunk 6]
```diff
@@ -327,13 +338,9 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
...
-	const allAvailablePlansProducts = useSelect(
-		( select ) =>
-			allAvailablePlans.map( ( plan ) =>
-				select( PLANS_STORE ).getPlanProduct( plan?.periodAgnosticSlug, billingPeriod )
-			),
-		[ allAvailablePlans, billingPeriod ]
-	);
+	const allAvailablePlansProducts: ( PlanProduct | undefined )[] = nonDefaultPaidPlanProduct
+		? [ defaultPaidPlanProduct, nonDefaultPaidPlanProduct, defaultFreePlanProduct ]
+		: [ defaultPaidPlanProduct, defaultFreePlanProduct ];
```
Comment (ciampo) [inline packages/launch/src/focused-launch/summary/index.tsx:337]: With these changes, the `allAvaiablePlans` and `allAvailablePlanProducts` arrays can technically go "out of sync".

We should think about a way of making sure that this doesn't happen — and so I though that one potential solution is to add one more check to https://github.com/Automattic/wp-calypso/blob/77f73e53b72bd1d5e074a0c61f1d8b7953807e21/packages/launch/src/focused-launch/summary/index.tsx#L422-L423

and change it to:

```typescript
typeof plan === 'undefined' ||
typeof allAvailablePlansProducts?.[ index ] === 'undefined' ||
plan.periodAgnosticSlug !== allAvailablePlansProducts?.[ index ]?.periodAgnosticSlug
```

Basically, the third check makes sure that the `plan` and the `planProduct` are in sync and refer to the same plan.

What do you think?

**Review by ciampo — COMMENTED** [review]
I left a few comments, although the changes seem to work well!

---

I also wanted to raise one more edge case that I came across while testing:

1. Open focused launch modal
2. In summary view, select the free subdomain
3. Go to Plan details view, and pick a _monthly billed_ plan
4. Go back to Summary view — everything looks as expected
5. Select the free plan in the Summary view — the Premium plan stays unchanged (monthly billed), even if it's not selected anymore
6. Go to Plan details and click the "Back" button, back to Summary view
7. The Free plan is selected, but now the plan above it is the _annually billed_ premium plan (and not the monthly)


https://user-images.githubusercontent.com/1083581/107975374-4f4c5400-6fb8-11eb-84d7-7c39bac0ee65.mp4



To me this is behaviour is acceptable (plus, it's very rare edge case), but I just wanted to document it in my review notes.

**Comment by ciampo** [comment]
Adding the "Hold" label because, as discussed in p1613476301450800/1613380064.426100-slack-C0KDTA48Y, we should wait for #48452 to be merged first, rebase, and then act on the points suggested by [this](https://github.com/Automattic/wp-calypso/pull/48452#discussion_r576304079) and [this](https://github.com/Automattic/wp-calypso/pull/48452#discussion_r576304197) comment

File: packages/data-stores/src/launch/selectors.ts (hunk 4 above)
Comment (alshakero) [reply packages/data-stores/src/launch/selectors.ts]: ~~There is no need for the check at all :D~~

Oops, wrong, there is a need. I address this.

File: packages/launch/src/focused-launch/summary/index.tsx (hunk 5 above)
Comment (alshakero) [reply packages/launch/src/focused-launch/summary/index.tsx]: This code is gone now ✨

File: packages/launch/src/focused-launch/summary/index.tsx (hunk 6 above)
Comment (alshakero) [reply packages/launch/src/focused-launch/summary/index.tsx:337]: I think they will never be out of sync. Two factors determine their order:

1. [These two](https://github.com/Automattic/wp-calypso/blob/bc18130ef55f6063d508160b636808a036590e1f/packages/launch/src/focused-launch/summary/index.tsx#L321-L322) synchronous lines. Because they control the ternary.
2. [These lines that use the same resolver](https://github.com/Automattic/wp-calypso/blob/970d6d6d049c8e8babb740edae0393e401268e22/packages/launch/src/hooks/use-plans.tsx#L32-L41) that won't resolve until all is fetched.

Nonetheless, I cleaned up that code a bit.

File: packages/data-stores/src/launch/actions.ts [hunk 7]
```diff
@@ -124,14 +143,35 @@ export const hideModalTitle = () =>
...
+type SetPlanActionParams = () =>
+	| {
+			readonly type: 'SET_PLAN_BILLING_PERIOD';
+			readonly billingPeriod: Plans.PlanBillingPeriod;
+	  }
+	| {
+			readonly type: 'SET_PLAN_PRODUCT_ID';
+			readonly planProductId: number | undefined;
+	  };
+
```
Comment (alshakero) [inline packages/data-stores/src/launch/actions.ts]: @sirreal Hi Jon! I couldn't figure out a way to type a generator action creator. Do you have a more elegant solution? This one isn't quite maintainable.

Comment (sirreal) [reply packages/data-stores/src/launch/actions.ts]: How about swapping `ReturnType` below for a new mapped type helper:

```ts
type ReturnOrGeneratorYieldUnion<T extends (...args: any) => any> = T extends (
	...args: any
) => infer Return
	? (Return extends Generator<infer T, infer U, any>
		? T | U
		: Return)
	: never;

export type LuanchAction = ReturnOrGeneratorYieldUnion<
  | typeof setPlanProductId
  | typeof setSiteTitle /* … */
>
```

I think you'll get the result you're looking for. [Here's a playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAShwFcBOA7A8kg4hFEkENgB7JATQEsIAbAEwB4AVKCAD2BxoGcoAKAOgH4kAc04AuKPhQgAlFAC8APknTl8qE1bsUXXgCgAkAL5DREqSD1ylUcigBmeWPGQpDAfl5xEqZmw7c2LgExEh0do5IGgA0tg5OAKqxFoqGBp5MAD5QCWkS3q4yhhK4AG54ANx6AMZEKJzAUPZERAq81so8AN5Q4lAARJz9UAC+kty19cAyVXqgkDltBagYQXiEJBTU9PMQRPa98AAKVFJHSEQ0CNXAAJI0UNm7+00tirOsYCSNkw2HwCczhcrjd7m17AgUDdyHUAFT-QEoc6Xa53Gg8QxgU5I4Go+4lBAAWwARk5spCaBB7HYIDQrFAuoZfo1yJwAGJICDQdTAJAIaD4bjElpUCBSWYGcgHHgAQlZHK5ckZBgMzKgxPIVCodmERzwMIe6n6AEEAHKmhLGgAyVtI-QlBhAlFoUAA+q7OPAAEKa7UoXX6y48DVanV6pAGmaGEZ6QycnwoBlpXYSfoAZQAogxXUcrWaczA0AARBIAYWztyL-WiaSxQJRoLpBjGgqgzKqI1mEKhwBhifdnuAPtD-vDBuDvrDgZoEgaEf9SrjLl8yoMKYGmezufzXtuNtupswOYzMFuxeraRDfoDEcu0fGbbqDQ7saAA).

Comment (alshakero) [reply packages/data-stores/src/launch/actions.ts]: Thank you so much! We could add this type to wp.data pacakge. Since it will be needed whenever someone needs a generator action.

Comment (alshakero) [reply packages/data-stores/src/launch/actions.ts]: I'm undecided between that above and this simpler solution:

```ts
export type LaunchAction =
	| ReturnType<
			| typeof setSiteTitle
			.
			.
			| typeof hideModalTitle
	  >
	| ReturnType< ReturnType< typeof setPlanProductId >[ 'next' ] >[ 'value' ]; // <--- this
```

File: packages/data-stores/src/launch/actions.ts [hunk 8]
```diff
@@ -124,7 +143,30 @@ export const hideModalTitle = () =>
...
+ * This works until one of the action creators is a generator and doesn't actually "Return" an action.
+ * This type helper allows for actions to be both functions and generators
+ */
+type ReturnOrGeneratorYieldUnion< T extends ( ...args: any ) => any > = T extends (
+	...args: any
+) => infer Return
+	? Return extends Generator< infer T, infer U, any >
+		? T | U
+		: Return
+	: never;
```
Comment (sirreal) [inline packages/data-stores/src/launch/actions.ts:174]: If this better describes action creators, we should move it with other helpers and use it in other store action creators as well.

https://github.com/Automattic/wp-calypso/blob/61acc957638fc7009632af4c0c058aa6bc0e82b7/packages/data-stores/src/mapped-types.ts#L31-L61

This is the type of thing wp-data should handle in its type system and I think a lot of the utilities we've developed here can move there when the library is typed. This is the kind of thing the system should take care of if we get it right and we shouldn't have to deal with them to build a soundly typed store 🙂

File: packages/data-stores/src/launch/actions.ts (hunk 7 above)
Comment (sirreal) [reply packages/data-stores/src/launch/actions.ts]: That's simpler here for a one-off case. I'd like to see a library-level solution for this that we could life to Gutenberg when it's ready like you suggest.

File: packages/data-stores/src/launch/actions.ts [hunk 9]
```diff
@@ -52,11 +55,27 @@ export const setDomainSearch = ( domainSearch: string ) =>
 		domainSearch,
 	} as const );
 
-export const setPlanProductId = ( planProductId: number | undefined ) =>
+const __setBillingPeriod = ( billingPeriod: Plans.PlanBillingPeriod ) =>
```
Comment (ciampo) [inline packages/data-stores/src/launch/actions.ts]: No need of prefixing with `__`, as this function is not exported anyway (plus, we may actually need it when working on #49253 and/or #50186)

File: packages/data-stores/src/launch/actions.ts [hunk 10]
```diff
@@ -52,11 +55,27 @@ export const setDomainSearch = ( domainSearch: string ) =>
...
 	( {
+		type: 'SET_PLAN_BILLING_PERIOD',
+		billingPeriod,
+	} as const );
+
+export const setPlanProductId = function* setPlanProductId( planProductId: number | undefined ) {
```
Comment (ciampo) [inline packages/data-stores/src/launch/actions.ts]: No need to repeat `setPlanProductId` twice. It can be rewritten to

```suggestion
export function* setPlanProductId( planProductId: number | undefined ) {
```

or, to keep consistency with the rest of the exported functions, to

```suggestion
export const setPlanProductId = function* ( planProductId: number | undefined ) {
```

File: packages/data-stores/src/plans/index.ts [hunk 11]
```diff
@@ -38,6 +38,7 @@ export {
 	TIMELESS_PLAN_PREMIUM,
 	TIMELESS_PLAN_BUSINESS,
 	TIMELESS_PLAN_ECOMMERCE,
+	FREE_PLAN_PRODUCT_ID,
```
Comment (ciampo) [inline packages/data-stores/src/plans/index.ts:41]: ~Looks like this export does not exist in the file~

Never mind this. Looks like my TypeScript server had a hiccup — all good :)

**Review by ciampo — APPROVED** [review]
LGTM 🚢 

(mobile e2e tests permitting)
